### PR TITLE
add new recipe major-mode-icons

### PR DESCRIPTION
**Pull request by stardiviner**
### Brief summary of what the package does

Show icon for major-mode on icon.

### Direct link to the package repository

https://github.com/stardiviner/major-mode-icons

### Your association with the package

Maintainer

### Relevant communications with the upstream package maintainer

[e.g., package.el compatibility changes that you have submitted. Write **None needed** if there is no problem.]

### Checklist

- [X] I've read [CONTRIBUTING.md](https://github.com/melpa/melpa/blob/master/CONTRIBUTING.md)
- [X] I've used [package-lint](https://github.com/purcell/package-lint) to check for packaging issues, and addressed its feedback
- [X] I've built and installed the package using the instructions in the [README](https://github.com/melpa/melpa/blob/master/README.md)

### CODE REVIEW

**Comment by syohex** [comment]
- Please conform Emacs Lisp package convention
  - missing some package headers(author, dependencies, license etc)
- Please don't use `/` separator, use `-`  instead of it(`--` can be used for private symbols)
- Please check too many byte-compile warnings

File: recipes/major-mode-icons [hunk 1]
```diff
@@ -0,0 +1,3 @@
+(major-mode-icons :fetcher github
+                  :repo "stardiviner/major-mode-icons"
+                  )
```
Comment (syohex) [inline recipes/major-mode-icons]: `icons/` directory is missing.

**Comment by syohex** [comment]
https://github.com/stardiviner/major-mode-icons/tree/master/icons

There are bug.xpm and Bug.xpm. Case insensitive file system like NTFS, HFS+ can see it as only one file.

**Comment by stardiviner** [comment]
@syohex All fixed. check it out the latest push.

**Comment by syohex** [comment]
- cl-lib functions are used but cl-lib is not loaded
- Don't use cl.el macro, use cl-lib macro instead 
- map-keys is introduced at Emacs 25.1 and you should load `map.el` for using it
- What is `activate` function ?
- There are two `Commentary` sections
- `Package-Requires` header position is wrong
-  Make no sense `;;; -*- coding: utf-8 -*-` at second line, it should be first line or buffer local variable section
- `This file is part of GNU Emacs.` Really ?
- `Copyright (C) 2016 Free Software Foundation, Inc` Really ?

Please check carefully before PR.

**Comment by stardiviner** [comment]
> Don't use cl.el macro, use cl-lib macro instead
What do you mean this?
Does `cl-some` has another name in `cl-lib`?

**Comment by syohex** [comment]
`case` is cl.el macro. Use `cl-case` instead of it. You can check it by byte-compiling with minimum configuration.

**Comment by stardiviner** [comment]
I updated my recipe to include `icons/` directory.

**Comment by purcell** [comment]
- Instead of `:files ("*.el" ("icons" "icons/*"))`, please just use `:files (:defaults "icons")`, which is equivalent.
- The `;;; Commentary:` section is misplaced. Please follow the standard formatting, as inserted by the `auto-insert` command.
- `cl-some` is autoloaded, so you don't need to require `cl-extra`
- The Emacs 25.1 dependency ensures that you'll already have a recent `cl-lib`, so you can remove the `(cl-lib "0.5")` dependency.
- It's possible to install `clojure-mode` but not `cider`, in which case your code will break. Same with `python-mode` and `pyvenv-mode`.
- The `post-command-hook` looks unnecessary. You can ask for the selected window at any time.

Overall, it looks like you should just modify `mode-line-format` in order to add an `(:eval ...)` section. What you're doing appears to be a hand-rolled version of that.

**Comment by purcell** [comment]
Also, you can get the preferred display name of a mode just by inspecting the `mode-name` variable in a buffer with that mode, so there's no need to hard-code a list of names, which might not match what the user wants.

**Comment by stardiviner** [comment]
I have changed my source code, waiting for merge. (maybe some places I missed, can figure out) Thanks for you guys review my code.

**Comment by syohex** [comment]
Please don't use `--` separator for customize variables. `--` means private function/variable but customize variables are public, not private.

**Comment by stardiviner** [comment]
@syohex Updated.

**Comment by syohex** [comment]
Merged

**Pull request merged**